assets/php/class-wgpb-block-grid-base.php

```diff
@@ -0,0 +1,340 @@
+
+		// Always query only IDs.
+		$query_args['fields'] = 'ids';
+
+		return $query_args;
+	}
+
+	/**
+	 * Set args specific to this block
+	 *
+	 * @param array $query_args Query args.
+	 */
+	abstract protected function set_block_query_args( &$query_args );
+
+	/**
+	 * Set categories query args.
+	 *
+	 * @param array $query_args Query args.
+	 */
+	protected function set_categories_query_args( &$query_args ) {
+		if ( ! empty( $this->attributes['categories'] ) ) {
+			$categories = array_map( 'absint', $this->attributes['categories'] );
+
+			$query_args['tax_query'][] = array(
+				'taxonomy'         => 'product_cat',
+				'terms'            => $categories,
+				'field'            => 'term_id',
+				'operator'         => 'all' === $this->attributes['catOperator'] ? 'AND' : 'IN',
+
+				/*
+				 * When cat_operator is AND, the children categories should be excluded,
+				 * as only products belonging to all the children categories would be selected.
+				 */
+				'include_children' => 'all' === $this->attributes['catOperator'] ? false : true,
+			);
+		}
+	}
+
+	/**
+	 * Run the query and return an array of product IDs
+	 *
+	 * @return array List of product IDs
+	 */
+	protected function get_products() {
+		// @todo Cache this
+		if ( 'product-top-rated' === $this->block_name ) {
+			add_filter( 'posts_clauses', array( __CLASS__, 'order_by_rating_post_clauses' ) );
+			$query = new WP_Query( $this->query_args );
+			remove_filter( 'posts_clauses', array( __CLASS__, 'order_by_rating_post_clauses' ) );
+		} else {
+			$query = new WP_Query( $this->query_args );
+		}
+
+		$results = wp_parse_id_list( $query->posts );
+
+		// Remove ordering query arguments which may have been added by get_catalog_ordering_args.
+		WC()->query->remove_ordering_args();
+
+		return $results;
+	}
+
+	/**
+	 * Render the Products block.
+	 *
+	 * @return string Rendered block type output.
+	 */
+	public function render() {
+		$products = $this->get_products();
+		$classes  = $this->get_classes();
+
+		$output = implode( '', array_map( array( $this, 'render_product' ), $products ) );
+
+		return sprintf( '<ul class="%s">%s</ul>', $classes, $output );
+	}
+
+	/**
+	 * Get the list of classes to apply to this block.
+	 *
+	 * @return string space-separated list of classes.
+	 */
+	public function get_classes() {
+		$classes = array(
+			'wc-block-grid',
+			"wp-block-{$this->block_name}",
+			"wc-block-{$this->block_name}",
+			"has-{$this->attributes['columns']}-columns",
+			'products',
+		);
+
+		if ( $this->attributes['rows'] > 1 ) {
+			$classes[] = 'has-multiple-rows';
+		}
+
+		if ( isset( $this->attributes['align'] ) ) {
+			$classes[] = "align{$this->attributes['align']}";
+		}
+
+		return implode( ' ', $classes );
+	}
+
+	/**
+	 * Render a single products.
+	 *
+	 * @param int $id Product ID.
+	 * @return string Rendered product output.
+	 */
+	public function render_product( $id ) {
+		global $product;
+		$product = wc_get_product( $id );
```

Comment:
Just a performance note, we should `_prime_post_caches` before rendering products so reading products is faster. Possibly in `get_products`